lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,300 @@
+        }),
+      };
+      getCache.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  getCache.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return addr.address === '127.0.0.1';
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
+  const parts = StringPrototypeSplit(str, ':');
+  let expansionIndex = -1;
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    if (part === '' && expansionIndex === -1 && parts.length < 8) {
+      expansionIndex = i;
+      continue;
+    }
+    parts[i] = StringPrototypePadStart(part, 4, '0');
+  }
+  if (parts.length < 8) {
+    ArrayPrototypeSplice(
+      parts,
+      expansionIndex,
+      1,
+      ...ArrayFrom({ length: 8 - parts.length + 1 }, () => '0000')
+    );
+  }
+  return ArrayPrototypeJoin(parts, ':');
+}
+
+/**
+ * @typedef CacheEntry
+ * @property resolvedHREF {Promise<string> | string}
+ * @property headers {Record<string, string>}
+ * @property body {Promise<Buffer> | Buffer}
+ */
+
+/**
+ * Only for GET requests, other requests would need new Map
+ * HTTP cache semantics keep diff caches
+ *
+ * Maps HREF to pending cache entry
+ * @type {Map<string, Promise<CacheEntry> | CacheEntry>}
+ */
+let getCache = new SafeMap();
+
+/**
+ * Fetches a location with a shared cache following redirects.
+ * Does not respect HTTP cache headers.
+ *
+ * This splits the header and body Promises so that things only needing
+ * headers don't need to wait on the body.
+ * @param {URL} parsed
+ * @returns {ReturnType<typeof fetchWithRedirects>}
+ */
+function fetch(parsed) {
+  const { href } = parsed;
+  const existing = getCache.get(href);
+  if (existing) {
+    return existing;
+  }
+  if (parsed.protocol === 'http:') {
+    return isLocalAddress(parsed.hostname).then((is) => {
+      if (is !== true) {
+        throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(`Cannot load non-local address ${href} from network using http: protocol, use https: instead`);
```

Comment:
It's annoying not being able to provide parent context here. Perhaps we could move this check to the resolver so that we can provide that?

Reply:
due to redirects this seems hard to hoist out but could be done potentially if we also return the final resolved location without doing a fetch in that case, but it makes the API complex.

Reply:
Do we support redirects to file URLs?

Reply:
Nope, intentionally we do not and that matches the web. Also, this is prevented due to the concern above about giving transitive and potentially compromised code paths access to powerful modules.

Reply:
In that case the ability to block remote addresses doesn't need to take into account redirects then per your previous comment?

Reply:
i'm not sure I understand, we only check the request/resolved URLs, not the redirects. we could add the redirects if desired

Reply:
I suggested a resolver check in https://github.com/nodejs/node/pull/36328#discussion_r687301098, which you argued woudl require supporting redirects, but if the only way to get to a remote address is through the resolver, and the a remote URL can never in turn redirect back into a local URL, then we know for sure that any resolution to a remote is a remote and can fully handle this check in the resolver and thus also having the parent context for a better error message?

Reply:
I'm just arguing in response to your comment in https://github.com/nodejs/node/pull/36328#discussion_r687345498.